vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -78,69 +132,176 @@
-                });
+                    if (date.year >= 0 && date.year <= 99) {
+                        // JS Date constructor converts years 0-99 to 1900-1999 so
+                        // this is needed to fix that when we want an explicit year.
+                        rawDate.setUTCFullYear(date.year);
+                    }
+                    let formattedDate = cleanString(rawDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
+                    if (date.year < 1000) {
+                        formattedDate = yearToFourDigitInFormattedDate(formattedDate);
+                    }
+                    return formattedDate;
+                };
+                datepicker.i18n.formatDate = tryCatchWrapper(formatDate);
 
-                datepicker.i18n.parseDate = tryCatchWrapper(function (dateString) {
+                /**
+                 * @param {string} dateString
+                 * @returns {DateHash|boolean|undefined}
+                 */
+                let _parseDate = function (dateString) {
                     dateString = cleanString(dateString);
 
-                    if (dateString.length == 0) {
+                    if (dateString.length === 0) {
                         return;
                     }
 
                     let match = dateString.match(datepicker.$connector.regex);
-                    if (match && match.length == 4) {
+                    if (match && match.length === 4) {
+                        /** @type {number} */
+                        let day, month, year
+                        /** @type {string} */
+                        let dayMatch, monthMatch, yearMatch;
+
                         for (let i = 1; i < 4; i++) {
-                            datepicker.$connector.parts[i-1].value = parseInt(match[i]);
+                            const part = parts[i-1];
+                            const matchValue = match[i];
+                            part.value = parseInt(matchValue);
+                            if (part === dayPart) {
+                                day = part.value;
+                                dayMatch = matchValue;
+                            } else if (part === monthPart) {
+                                month = part.value - 1;
+                                monthMatch = matchValue;
+                            } else if (part === yearPart) {
+                                year = part.value;
+                                yearMatch = matchValue;
+                            }
                         }
-                        return {
-                            day: datepicker.$connector.dayPart.value,
-                            month: datepicker.$connector.monthPart.value - 1,
-                            year: datepicker.$connector.yearPart.value
-                        };
-                    }  else {
+                        return {day, month, year, dayMatch, monthMatch, yearMatch};
```

Comment:
Added `*Match` properties here with the original string matches of each part in the given date string. These are used in `indexOfYearForFormattedDate()` to check if a day or month is specified with 1 or 2 digits (those are also zero padded in some locales like UK).